redis/src/sentinel.rs

```diff
@@ -0,0 +1,558 @@
+            .async_find_valid_replica_addresses(service_name)
+            .await?;
+        if !addresses.is_empty() {
+            self.replica_start_index = (self.replica_start_index + 1) % addresses.len();
+        }
+        try_connect_to_first_replica(&addresses, self.replica_start_index)
+    }
+}
+
+// non-async methods
+impl Sentinel {
+    /// Creates a Sentinel client performing some basic
+    /// checks on the URLs that might make the operation fail.
+    pub fn build<T: IntoConnectionInfo>(params: Vec<T>) -> RedisResult<Sentinel> {
+        Ok(Sentinel {
+            sentinels_connection_info: params
+                .into_iter()
+                .map(|p| p.into_connection_info())
+                .collect::<RedisResult<Vec<ConnectionInfo>>>()?,
+            replica_start_index: random_replica_index(1000000),
+        })
+    }
+
+    fn try_all_sentinels<F, T>(&self, f: F) -> RedisResult<T>
+    where
+        F: Fn(&ConnectionInfo) -> RedisResult<T>,
+    {
+        let mut last_err = None;
+        for connection_info in self.sentinels_connection_info.iter() {
+            match f(connection_info) {
+                Ok(result) => {
+                    return Ok(result);
+                }
+                Err(err) => {
+                    last_err = Some(err);
+                }
+            }
+        }
+
+        // We can unwrap here because we know there is at least one connection info.
+        // Maybe we should have a more specific error for this situation?
+        Err(last_err.expect("There should be at least one connection info"))
+    }
+
+    fn get_sentinel_masters(&self) -> RedisResult<Vec<HashMap<String, String>>> {
+        self.try_all_sentinels(try_sentinel_masters)
+    }
+
+    fn get_sentinel_replicas(
+        &self,
+        service_name: &str,
+    ) -> RedisResult<Vec<HashMap<String, String>>> {
+        self.try_all_sentinels(|conn_info| try_sentinel_replicas(conn_info, service_name))
+    }
+
+    fn find_master_address(&self, service_name: &str) -> RedisResult<(String, u16)> {
+        let masters = self.get_sentinel_masters()?;
+        get_first_valid_master(masters, service_name)
+    }
+
+    fn find_valid_replica_addresses(&self, service_name: &str) -> RedisResult<Vec<(String, u16)>> {
+        let replicas = self.get_sentinel_replicas(service_name)?;
+        Ok(get_valid_replicas_addresses(replicas))
+    }
+
+    /// Determines the masters address for the given name, and returns a client for that
+    /// master.
+    ///
+    /// To determine the master we ask the sentinels one by one until one of
+    /// then correctly responds to the SENTINEL MASTERS command, and then we check to
+    /// see if there is a master with the given name; if there is, we create a
+    /// client with that address (ip and port).
+    pub fn master_for(&self, service_name: &str) -> RedisResult<Client> {
+        let address = self.find_master_address(service_name)?;
+        Client::open(address)
```

Comment:
minor: why a client, and not connection info?

Reply:
I just hadn't thought about this possibility, you think it would be better? Is a connection info useful for anything else than creating a client? :thinking:

Reply:
well, for starters, connection info can be enriched to add TLS :)
not a big deal, it's my preference to return the basic data.